onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java

```diff
@@ -111,33 +111,34 @@ public TaskContextImpl(int startId) {
         public void taskComplete() {
             //Log.d(TAG, "Task complete: " + mStartId);
             // If we have notifications, then we can't stop ourselves.
-            if (mNotifications.isEmpty()) {
+            //TODO: Is there actually a reason to not stop ourselves if we are no longer storing notifications?
```

Comment:
I'm wondering if the "right" way to ensure that all tasks execute to completion is to maintain a data structure that contains all the [`Future` objects](http://developer.android.com/reference/java/util/concurrent/Future.html) returned each time the `mThreadPool.submit()` is called to submit the Runnable for execution.  Then, instead of:

`if (mNotifications.isEmpty()) {`

... we'd do:

~~~
boolean allDone = true;
for (Future f : futures) {
    if (!f.isDone()) {
        allDone = false;
        break;
    }
}

if (allDone) {
    stopSelfResult(mStartId);
}
~~~